### ISSUE TITLE
Extend Sitemap functionality to support Portals

### ISSUE THREAD

**Issue by amoeba** [NCEAS/metacat#1432]
(associated MetacatUI issue is https://github.com/NCEAS/metacatui/issues/1302)

It would be really great if Portals were listed in the sitemaps to maximize the chance of them being crawled by search engines. The sitemap functionality currently _should list_ portal documents in sitesmaps but the `<url>` entry in the sitemap for that portal will use the `/view/$PID` URL and not the configured Portal URL space the user configures with MetacatUI, such as `/portal/$LABEL`.

I don't think there's a clear best way to go about this so here are some options:

1. Add in configuration for a fixed set of URLs to always append to the sitemaps. This would require Tomcat restarts after any reconfiguration but is nice because it doesn't require us to refactor the Sitemap class.
2. Refactor the Sitemap class and associated configuration so a URL prefix can be set for both metadata docs and Portal docs the Sitemap class can have a way to dynamically choose the right URL prefix for each Object it's putting in the sitemap files. Configuring a Portal prefix could be optional.

Let's discuss this on March 12th's developer call.

**Implementation**

We went with 2 (above): Two optional Metacat properties are included which control special behavior in the Sitemap class when set:

```
sitemap.entry.portal.base=/metacatui/portals
sitemap.entry.portal.formats=https://purl.dataone.org/portals-1.0.0
```

These are now version control defaults so portal links work by default. If absent, the value set in `sitemap.entry.base` is used which should also still work fine. When processing entries to go into the sitemap, the `Sitemap` class checks for the format ID against the value(s) set in `sitemap.entry.portal.formats` to detect portal docs and links to them via the value set in `sitemap.entry.portal.base` is so. The link attempts to link to the portal document if it has a series ID but will fall back to the PID if the series ID is absent. Non-portal docs are still linked to by their PID and not series ID.

**Comment by laurenwalker** [comment]
Hey @amoeba - I was wondering how much time you estimate this would take? It would be great if these could start being indexed by search engines soon

**Comment by amoeba** [comment]
Hey @laurenwalker, I think I can wrap the code up by the end of the week and I think I'd go with option number 2 above. At that point we'd need to coordinate with @taojing2002 on prepping a 2.13 release of Metacat and getting it deployed. I'd guess that might take ~1month if we fast track it at this point.

Related, Googlebot seems to parse our sitemaps correctly but also totally ignore them and crawls our web pages directly so if we have public links to portals from somewhere in the search UI I'd bet they'd get picked up.

**Comment by laurenwalker** [comment]
> Related, Googlebot seems to parse our sitemaps correctly but also totally ignore them and crawls our web pages directly so if we have public links to portals from somewhere in the search UI I'd bet they'd get picked up.

Unfortunately we only have links to portals when you're logged in and viewing your portal list, but I have been wanting to make a quick /portals view sometime soon that would simply list all public portals.

**Comment by amoeba** [comment]
Done and tested in https://github.com/NCEAS/metacat/commit/721f51d85bae76dc8c6a78b7910e5c82c57671e0.

**Comment by laurenwalker** [comment]
Awesome, thanks! 

However it looks like you set the URL base for the portals incorrectly - portals should be plural - `sitemap.entry.portal.base=/metacatui/portals`

Also, we can remove `https://purl.dataone.org/collections-1.0.0` as a portals formatID

It also looks like the URLs are generated with the portal pids? The portal URLs use the `label` in the URL, so we should use the label in the sitemap as well.

**Comment by amoeba** [comment]
I fixed the plural thing in a followup commit. 

Re:

> Also, we can remove https://purl.dataone.org/collections-1.0.0 as a portals formatID

Will do.

> It also looks like the URLs are generated with the portal pids? The portal URLs use the label in the URL, so we should use the label in the sitemap as well.

Ah, yeah. I'm a little too unfamiliar with portals I guess. We don't readily have access to the portal document or the Solr document for the record during sitemap generation. Because sitemap generation needs to scale to millions of records, I'm hesitant to take much of a performance hit. Can we work around the limitation in MetacatUI by handling entry in a `/portals` URL from its PID and updating the URL to the label after it's been found?

**Comment by laurenwalker** [comment]
Could we use the seriesid in the sitemap? That would be a nice compromise between the label and the pid

**Comment by amoeba** [comment]
Yeah, we have full access to the database so we can snag anything from sysmeta, for example.

**Comment by laurenwalker** [comment]
Okay, let's use the seriesId in the sitemap if we can't use the label and if that is just as scalable as using the pid

**Comment by amoeba** [comment]
Yep, will do. Thanks looking and catching this stuff.

**Comment by amoeba** [comment]
Alright: As of https://github.com/NCEAS/metacat/commit/1bb93db4ec57c0130d70d07ae18ca52c6c6fe713, the sitemap routine attempts to link to the portal by series ID if present, and falls back to the PID since portals docs don't necessarily have SIDs. Current implementation is described in an edit to the OP.

**Comment by amoeba** [comment]
Moved from 2.13.0 milestone to 2.14.0 because the changes never saw a 2.13 series release. We've merged all changes on the 2.13 branch into the 2.14 branch so this and other changes will go into the first 2.14 series release.
